### PR TITLE
Cura 10220 support towers reenabled

### PR DESCRIPTION
**Pull request by casperlamboo**
Merge this along with front-end PR: https://github.com/Ultimaker/Cura/pull/14515

### CODE REVIEW

File: src/support.cpp [hunk 1]
```diff
@@ -891,9 +891,10 @@ void AreaSupport::generateSupportAreasForMesh(SliceDataStorage& storage,
                                            Polygons larger_area_below; // the areas in the layer below that protrude beyond the area of the current layer
                                            if (layer_idx > 1)
                                            {
-                                               larger_area_below = mesh.layers[layer_idx - 1].getOutlines().difference(mesh.layers[layer_idx].getOutlines());
+                                               // shrink a little so that areas that only protrude very slightly are ignored
+                                               larger_area_below = mesh.layers[layer_idx - 1].getOutlines().difference(mesh.layers[layer_idx].getOutlines()).offset(-layer_thickness / 10);
 
-                                               if (!larger_area_below.empty())
+                                               if (larger_area_below.size())
```
Comment (jellespijker) [inline src/support.cpp:897]:
```suggestion
                                               if (! larger_area_below.empty())
```

File: src/support.cpp [hunk 2]
```diff
@@ -891,9 +891,10 @@ void AreaSupport::generateSupportAreasForMesh(SliceDataStorage& storage,
                                            Polygons larger_area_below; // the areas in the layer below that protrude beyond the area of the current layer
                                            if (layer_idx > 1)
                                            {
-                                               larger_area_below = mesh.layers[layer_idx - 1].getOutlines().difference(mesh.layers[layer_idx].getOutlines());
+                                               // shrink a little so that areas that only protrude very slightly are ignored
+                                               larger_area_below = mesh.layers[layer_idx - 1].getOutlines().difference(mesh.layers[layer_idx].getOutlines()).offset(-layer_thickness / 10);
```
Comment (jellespijker) [inline src/support.cpp:895]: magic number `10`

File: src/support.cpp [hunk 3]
```diff
@@ -1060,12 +1043,10 @@ void AreaSupport::generateSupportAreasForMesh(SliceDataStorage& storage,
...
                     if (layer_idx < layer_count - tower_top_layer_count && layer_idx >= tower_top_layer_count + bottom_empty_layer_count)
                     {
-                        const Polygons& layer_below = xy_disallowed_per_layer[layer_idx - tower_top_layer_count - bottom_empty_layer_count];
-                        const Polygons& layer_above = support_areas[layer_idx + tower_top_layer_count];
+                        const Polygons& layer_below = storage.getLayerOutlines(layer_idx - tower_top_layer_count - bottom_empty_layer_count, no_support, no_prime_tower);
                         const Point middle = AABB(poly).getMiddle();
-                        const bool has_support_above = layer_above.inside(middle);
                         const bool has_model_below = layer_below.inside(middle);
-                        if (has_support_above && ! has_model_below)
+                        if (!has_model_below)
```
Comment (jellespijker) [inline src/support.cpp:1049]: code-style

```suggestion
                        if (! has_model_below)
```

File: src/support.cpp [hunk 4]
```diff
@@ -1314,16 +1295,8 @@ std::pair<Polygons, Polygons> AreaSupport::computeBasicAndFullOverhang(const Sli
...
-        const auto nozzle_diameter = mesh.settings.get<coord_t>("machine_nozzle_size");
-        const coord_t min_circumference = nozzle_diameter * M_PI;
-        const double min_area = INT2MM2((nozzle_diameter * nozzle_diameter) / 4 * M_PI);
-        constexpr bool remove_holes = true;
-        basic_overhang.removeSmallAreaCircumference(min_area, min_circumference, remove_holes);
-    }
-
-    Polygons overhang_extended = basic_overhang.offset(max_dist_from_lower_layer + MM2INT(0.1)); // +0.1mm for easier joining with support from layer above
-    Polygons full_overhang = overhang_extended.intersection(supportLayer_supportee);
+    Polygons overhang_extented = basic_overhang.offset(max_dist_from_lower_layer + MM2INT(0.1)); // +0.1mm for easier joining with support from layer above
```
Comment (jellespijker) [inline src/support.cpp:1298]: Can you remove the ancient commented out code above, even though it isn't part of your changes?

File: src/support.cpp [hunk 5]
```diff
@@ -1399,12 +1372,24 @@ void AreaSupport::handleTowers(const Settings& settings, Polygons& supportLayer_
     // make tower roofs
     const coord_t layer_thickness = settings.get<coord_t>("layer_height");
     const AngleRadians tower_roof_angle = settings.get<AngleRadians>("support_tower_roof_angle");
-    const double tan_tower_roof_angle = tan(tower_roof_angle);
-    const coord_t tower_roof_expansion_distance = layer_thickness / tan_tower_roof_angle;
     const coord_t tower_diameter = settings.get<coord_t>("support_tower_diameter");
+    const coord_t support_line_width = settings.get<coord_t>("support_line_width");
```
Comment (jellespijker) [inline src/support.cpp:1376]: type duplication, it is already specified in the `get` template argument
```suggestion
    const auto support_line_width = settings.get<coord_t>("support_line_width");
```

File: src/support.cpp [hunk 6]
```diff
@@ -1399,12 +1372,24 @@ void AreaSupport::handleTowers(const Settings& settings, Polygons& supportLayer_
...
+        tower_roof_expansion_distance = layer_thickness / tan_tower_roof_angle;
+    }
+
     for (Polygons& tower_roof: tower_roofs)
     {
-        if (tower_roof.size() == 0)
+        if (tower_roof.empty())
         {
             continue;
         }
```
Comment (jellespijker) [inline src/support.cpp:1395]: consider using ranges filter
```suggestion
    for (Polygons& tower_roof: tower_roofs | ranges::views::filter([](const auto& poly){ return ! poly.empty(); }))
    {
```

File: src/support.cpp [hunk 7]
```diff
@@ -1413,7 +1398,30 @@ void AreaSupport::handleTowers(const Settings& settings, Polygons& supportLayer_
 
         if (tower_roof[0].area() < tower_diameter * tower_diameter)
         {
-            tower_roof = tower_roof.offset(tower_roof_expansion_distance);
+            const bool no_support = false;
+            const bool no_prime_tower = false;
```
Comment (jellespijker) [inline src/support.cpp:1402]: use `constexpr`

```suggestion
            constexpr bool no_support = false;
            constexpr bool no_prime_tower = false;
```

File: src/support.cpp [hunk 8]
```diff
@@ -1413,7 +1398,30 @@ void AreaSupport::handleTowers(const Settings& settings, Polygons& supportLayer_
...
+            Polygons model_outline = storage.getLayerOutlines(layer_idx, no_support, no_prime_tower);
+
+            // Rather than offsetting the tower with tower_roof_expansion_distance we do this step wise to achieve two things
+            // - prevent support from folding around the model
+            // - provide method to early out the offsetting procedure when the desired area is reached
+            const coord_t offset_per_step = support_line_width / 2;
+            for (coord_t offset_cumulative = 0; offset_cumulative <= tower_roof_expansion_distance; offset_cumulative += offset_per_step)
```
Comment (jellespijker) [inline src/support.cpp:1409]: does this have to be such an ancient for loop?
Feels like this can be done with a range

**Review by jellespijker — CHANGES_REQUESTED** [review]

File: src/support.cpp (hunk 2 above)
Comment (casperlamboo) [reply src/support.cpp:895]: ✨

File: src/support.cpp (hunk 6 above)
Comment (casperlamboo) [reply src/support.cpp]: https://github.com/Ultimaker/CuraEngine/pull/1824/commits/10194058e601e86d28c240db8db59691f0f1674c

**Pull request merged**